## TECHIN 513 HW1 – Modifying Signals

Ghea C Suyono

### Assignment 1: Working with Sound Files

In [23]:
# Import relevant libraries

import numpy as np
import matplotlib.pyplot as plt
from scipy.io import wavfile as wav
import simpleaudio as sa
import scipy.signal as sig

#### Part A

##### Use Scipy.io to read audio file
Note: 
- fs1: sampling frequency 1
- x1: audio vector


In [6]:
#import train32 file
fs1, x1 = wav.read('sounds/train32.wav')
#print sample rate and vector
print('Sampling rate:', fs1, 'Audio vector', x1)
x1.shape


Sampling rate: 32000 Audio vector [-901 -943 -646 ...  330  169   45]


(50313,)

The data above shows that:
1. The sampling rate of train32 file is 32kHz 
2. The shape of audio data array is 1D array (mono) with 50,313 samples.

In [36]:
#Perform time scaling on the audio signal
#Define the time scaling factor
# a: scaling parameter. This has to be a decimal value for as_integer_ratio

def timescale(x, fs, a):
    # 1. Convert the scaling factor to a ratio of integers
    ratio = (np.double(a)).as_integer_ratio()
    
    # 2. Define the numerator and denominator of the ratio
    numerator = ratio[0]
    denominator = ratio[1]
    
    # 3. Resample the audio signal use scipy.signal (sig)
    #Convert input signal (x) to a floating point number to ensure data type is consistent
    # y: resampled audio signal
    y = sig.resample_poly([float(i) for i in x], denominator, numerator)
    
    # 4. Define the new audio vector -> np.arange(start, stop, tsample)
    t = np.arange(0, len(y), 1) * (1/fs)
    
    # 5. Returns t: time samples vector corresponding to y: scaled signal
    return y, t


In [37]:
#Perform time scaling on the train32 audio signal
a = 1.5

scaled_signal, scaled_time_vector = timescale(x1, fs1, a)
print('Scaled signal:', scaled_signal, 'Scaled time vector:', scaled_time_vector)

Scaled signal: [-777.06971284 -867.64903067 -248.37940892 ...  546.0978636   330.86783035
   98.78339607] Scaled time vector: [0.00000000e+00 3.12500000e-05 6.25000000e-05 ... 1.04809375e+00
 1.04812500e+00 1.04815625e+00]


#### Part B

##### Create two new versions of the file with different sampling rates.
- fs2: 1/2fs1
- fs3: 1.5fs1

In [47]:
# 1. Define new sampling rates
fs2 = int(fs1 / 2) # 1/2 of the original sampling rate
fs3 = int(1.5 * fs1)  # 1.5 times the original sampling rate

# 2. Resample the audio signal using scipy.signal (sig)
# y2: resampled audio signal with fs2
# y3: resampled audio signal with fs3

y2 = sig.resample_poly([float(i) for i in x1], 2, 1)
y3 = sig.resample_poly([float(i) for i in x1], 2, 3)

# 3. Write the new audio file using scipy.io.wavfile.write
wav.write('sounds/train32_2.wav', fs2, y2.astype(np.int16))
wav.write('sounds/train32_3.wav', fs3, y3.astype(np.int16)) # Convert to 16-bit integers



#### Part C

##### Read 3 different version of audio file using simple audio

In [49]:
def play_audio(file_path):
    wave_obj = sa.WaveObject.from_wave_file(file_path)
    play_obj = wave_obj.play()
    play_obj.wait_done()

# Play the original file
play_audio('sounds/train32.wav')

# Play the modified files
play_audio('sounds/train32_2.wav')
play_audio('sounds/train32_3.wav')

#### Report Discussion

- Fs2 ('train32_2.wav'): Slowing down the audio by halfing the sample rate makes the audio sounds unclear.
- Fs3 ('train32_3.wav'): Speeding up the audio by multiplying 1.5 to the original sample rate resulted in higher pitch audio sound.